lib/entities/quote/DutchV2Quote.ts

```diff
@@ -0,0 +1,221 @@
+import { PermitTransferFromData } from '@uniswap/permit2-sdk';
+import { TradeType } from '@uniswap/sdk-core';
+import { V2DutchOrder, V2DutchOrderBuilder, V2DutchOrderInfoJSON } from '@uniswap/uniswapx-sdk';
+import { BigNumber, ethers } from 'ethers';
+
+import { IQuote, LogJSON } from '.';
+import { DutchV2Request } from '..';
+import { BPS, frontendAndUraEnablePortion, RoutingType } from '../../constants';
+import { Portion } from '../../fetchers/PortionFetcher';
+import { generateRandomNonce } from '../../util/nonce';
+import { currentTimestampInMs, timestampInMstoSeconds } from '../../util/time';
+import { DutchQuote as DutchV1Quote, getPortionAdjustedOutputs } from './DutchQuote';
+
+// TODO: replace with real cosigner when deployed
+export const LABS_COSIGNER = '0x0000000000000000000000000000000000000000';
+
+// JSON format of a DutchV2Quote, to be returned by the API
+export type DutchV2QuoteDataJSON = {
+  orderInfo: V2DutchOrderInfoJSON;
+  quoteId: string;
+  requestId: string;
+  encodedOrder: string;
+  orderHash: string;
+  deadlineBufferSecs: number;
+  slippageTolerance: string;
+  permitData: PermitTransferFromData;
+  portionBips?: number;
+  portionAmount?: string;
+  portionRecipient?: string;
+};
+
+type DutchV2QuoteConstructorArgs = {
+  createdAtMs?: string;
+  request: DutchV2Request;
+  chainId: number;
+  requestId: string;
+  quoteId: string;
+  tokenIn: string;
+  tokenOut: string;
+  amountInStart: BigNumber;
+  amountInEnd: BigNumber;
+  amountOutStart: BigNumber;
+  amountOutEnd: BigNumber;
+  swapper: string;
+  nonce?: string;
+  portion?: Portion;
+};
+
+export class DutchV2Quote implements IQuote {
+  public readonly routingType: RoutingType.DUTCH_V2 = RoutingType.DUTCH_V2;
+
+  public readonly request: DutchV2Request;
+  public readonly createdAtMs: string;
+  public readonly chainId: number;
+  public readonly requestId: string;
+  public readonly quoteId: string;
+  public readonly tokenIn: string;
+  public readonly tokenOut: string;
+  public readonly amountInStart: BigNumber;
+  public readonly amountInEnd: BigNumber;
+  public readonly amountOutStart: BigNumber;
+  public readonly amountOutEnd: BigNumber;
+  public readonly swapper: string;
+  public readonly nonce?: string;
+  public readonly portion?: Portion;
+
+  // build a v2 quote from a v1 quote
+  public static fromV1Quote(request: DutchV2Request, quote: DutchV1Quote): DutchV2Quote {
+    return new DutchV2Quote({
+      ...quote,
+      request,
+    });
+  }
+
+  private constructor(args: DutchV2QuoteConstructorArgs) {
+    Object.assign(this, args, {
+      createdAtMs: args.createdAtMs || currentTimestampInMs(),
+    });
+    this.routingType = RoutingType.DUTCH_V2;
```

Comment:
don't think you need to assign here

Reply:
oh true

Reply:
ok strangely we do - when I remove it, the type is `DUTCH_LIMIT`. I think the inheritance overrides it

Reply:
interesting lol